build/Build.cs

```diff
@@ -0,0 +1,616 @@
+using Nuke.Common.Tools.OctoVersion;
+using Nuke.Common.Tools.SignTool;
+using Nuke.Common.Utilities.Collections;
+using Serilog;
+using static Nuke.Common.IO.FileSystemTasks;
+using static Nuke.Common.Tools.DotNet.DotNetTasks;
+using static Nuke.Common.Tools.Git.GitTasks;
+using static Nuke.Common.Tools.AzureSignTool.AzureSignToolTasks;
+
+namespace Calamari.Build
+{
+    class Build : NukeBuild
+    {
+        const string LinuxRuntime = "linux-x64";
+        const string WindowsRuntime = "win-x64";
+        const string CiBranchNameEnvVariable = "OCTOVERSION_CurrentBranch";
+        const string RootProjectName = "Calamari";
+
+        readonly string[] TimestampUrls =
+        {
+            "http://timestamp.digicert.com?alg=sha256",
+            "http://timestamp.comodoca.com"
+        };
+
+        [Parameter("Configuration to build - Default is 'Debug' (local) or 'Release' (server)")]
+        readonly Configuration Configuration =
+            IsLocalBuild ? Configuration.Debug : Configuration.Release;
+
+        [Solution] [Required] readonly Solution? Solution;
+
+        [Parameter("Whether to auto-detect the branch name - this is okay for a local build, "
+                   + "but should not be used under CI.")]
+        readonly bool AutoDetectBranch = IsLocalBuild;
+
+        [Parameter("Branch name for Calamari to use to calculate the version number. Can be set "
+                   + "via the environment variable " + CiBranchNameEnvVariable + ".", Name = CiBranchNameEnvVariable)]
+        readonly string? BranchName;
+
+        [Parameter("Run packing step in parallel")] 
+        readonly bool PackInParallel;
+
+        [Parameter("Build verbosity")] 
+        readonly string BuildVerbosity = "Minimal";
+
+        [Parameter("Sign Binaries")] 
+        readonly bool SignBinaries;
+
+        [Parameter("Append Timestamp")] 
+        readonly bool AppendTimestamp;
+
+        [Parameter("Set Calamari Version on OctopusServer")] 
+        readonly bool SetOctopusServerVersion;
+
+        [Parameter("AzureKeyVaultUrl")] 
+        readonly string? AzureKeyVaultUrl;
+
+        [Parameter("AzureKeyVaultAppId")] 
+        readonly string? AzureKeyVaultAppId;
+
+        [Parameter("AzureKeyVaultAppSecret")] 
+        readonly string? AzureKeyVaultAppSecret;
+
+        [Parameter("AzureKeyVaultCertificateName")] 
+        readonly string? AzureKeyVaultCertificateName;
+
+        [Parameter("SigningCertificatePath")]
+        readonly string SigningCertificatePath = "./certificates/OctopusDevelopment.pfx";
+
+        [Parameter("SigningCertificatePassword")] 
+        readonly string SigningCertificatePassword = "Password01!";
+
+        [Required]
+        [OctoVersion(BranchParameter = nameof(BranchName), AutoDetectBranchParameter = nameof(AutoDetectBranch))]
+        readonly OctoVersionInfo? OctoVersionInfo;
+
+        public Build()
+        {
+            NugetVersion = new Lazy<string>(() => AppendTimestamp
```

Comment:
The formatting here is a bit wonky, and I'm not sure I quite understand why it's a Lazy in this situation.

Reply:
I think I did that because during construction the OctoVersionInfo property is not set? Or maybe I just thought it wouldn't be set... I'll try making it not a lazy and see how it goes.

Reply:
> I've had a read through it and it's looking like it's on the right track, but I'd be keen to work with you a bit tomorrow (If that works for you?) and we can try and refactor this a bit and pair review it together. Would you be down for that @scottmerchant ?

Yes that would be great, thanks!

Reply:
Checking it in the debugger - in the constructor the OctoVersionInfo is still null so I need to delay the evaluation until later.